Newsletter: Avoid these AI coding mistakes (#12809)
PostHog/posthog.com: contents/newsletter/ai-coding-mistakes.md

- AI can go and [make changes to parts of your apps you don’t expect](/blog/devtools-advice-agent-llm). Radically changing the UI might be fine in a small prototype but it can ruin a lot of things in a big app like ours.

- Tests, linting, and type checking increase in importance as they help protect against AI making changes with unintended consequences. 

- Be specific with your prompts. In a larger codebase, individual files and features can take over your context window. Vaguely asking “make it better” will leave the agent confused and ineffective.

As one of our product engineers, Paul, said in [his blog on how he uses LLMs](https://pauldambra.dev/2025/07/how-i-use-llms.html):

> They're not actually very good software engineers… particularly since most of the data they've ingested about software engineering is "blogs on how to start something from scratch". So, if that's not the task. Then I find it often harder to prompt an LLM to do than to do it myself

> ## 2. Not having rules and guardrails

Reviewer: DeskHog provides a pretty meaty example:

- microcontrollers have specific pins that map to certain devices and they vary from board to board. In the absence of a concrete pin definition, the LLM just makes guesses. With dozens of pin combinations, they're always wrong. So it's important to provide either a pinout document or source code related to a specific board
- relatedly, different devices ON a board have different needs for initialization and setup; providing a markdown file that captures documentation and some example code demonstrating a proper init sequence is the difference between quick prototyping and being miserable
- as libraries evolve, their function signatures and other details can change subtly; often you'll have a copy of the source code right there and waiting locally, and you can point the agent in that direction to course-correct
- It can also be very helpful to have a Deep Research product, like ChatGPT and Claude offer, to go out and scour forums about the particular technologies you're targeting and gather a markdown file with the most important, up-to-date details. Providing this to the agent keeps it from guessing because it's got the benefit of first-hand information from real people

is that a helpful rundown? Anything I can go deeper about?

---
title: 7 mistakes to avoid when coding with AI
date: 2025-09-16
author:
 - ian-vanagas
tags:
 - AI
---

No matter how much hype (or money) AI code editors get, engineers are still asking: Do they actually make us better?

At PostHog, the response is mixed. Some find autocomplete great, but need to rip out anything beyond that. Others find they are a huge speedup, but only in specific circumstances.

What is clear for everyone is that coding with AI is a skill. It takes practice and a lot of mistakes to get good at it.

> We know this because we've made a bunch of mistakes over the past few years working with these tools, building AI-powered features, and working with companies in the space. We're sharing them here to help you avoid them.

Reviewer 2: ```suggestion
We know this because we've made a bunch of mistakes over the past few years working with these tools, [building AI-powered features](/newsletter/building-ai-features), and working with companies in the space. Here's what we've learned.
```

Like AI, the rise in importance of these tools isn’t expected to slow down either. As [Gergely from Pragmatic Engineer](https://newsletter.pragmaticengineer.com/p/software-engineering-with-llms-in-2025) says:

> **Google is preparing for 10x more code to be shipped.** A former Google Site Reliability Engineer (SRE) told me:
> “What I’m hearing from SRE friends is that they are preparing for 10x the lines of code making their way into production.”
> If any company has data on the likely impact of AI tools, it’s Google. 10x as much code generated will likely also mean 10x more:
> - Code review
> - Deployments
> - Feature flags
> - Source control footprint
> - … and, perhaps, even bugs and outages, if not handled with care

## 5. Trying to use AI on something you know it’s not good at

> Claude Code writing Rust is a `while` loop that accelerates climate change - [Nick Best](/community/profiles/33511), Team Ingestion Product Engineer

Using AI most effectively requires piling up a collection of examples of situations where AI is and isn’t useful. You can waste a lot of time and energy repeatedly asking it to do something you know it can’t do.

> ![Insanity](https://res.cloudinary.com/dmukukwp6/image/upload/q_auto,f_auto/0_hoo_T68_VL_Ya_Z_gqsl_b26e1a6a0e.jpg)

Author: One to Hogify
